### PR TITLE
[interoperator] Error handling

### PR DESCRIPTION
**Pull request by vivekzhere**
- Define interoperator errors
- Pass interoperator errors to controllers
- Delete binding if instance not found
- Controllers dont fail delete if some resource is not found
- Add testcases

### CODE REVIEW

**Comment by subhankarc** [comment]
@vivekzhere run the pipeline and keep it ready. Will merge early next week.

**Comment by vivekzhere** [comment]
Dev Pipeline build 781 green

**Comment by vivekzhere** [comment]
> 
> 
> @vivekzhere run the pipeline and keep it ready. Will merge early next week.

Dev Pipeline build 781 green

**Comment by subhankarc** [comment]
@amitmalav please review this.

@vivekzhere is there a need to test it with large data size? Or are we good there?

File: interoperator/pkg/controller/sfservicebinding/sfservicebinding_controller.go [hunk 1]
```diff
@@ -322,11 +324,17 @@ func (r *ReconcileSFServiceBinding) updateUnbindStatus(targetClient client.Clien
 	bindingID := binding.GetName()
 	namespace := binding.GetNamespace()
 	computedStatus, err := r.resourceManager.ComputeStatus(r, targetClient, instanceID, bindingID, serviceID, planID, osbv1alpha1.BindAction, namespace)
-	if err != nil {
+	if err != nil && !errors.IsNotFound(err) {
 		log.Error(err, "ComputeStatus failed for unbind", "binding", bindingID)
 		return err
 	}
 
+	if errors.IsNotFound(err) && computedStatus == nil {
```
Comment (amitmalav) [inline interoperator/pkg/controller/sfservicebinding/sfservicebinding_controller.go]: What does this block of code do? Shouldn't we delete the binding crd if operator bind crd is not present?

File: interoperator/pkg/errors/errors.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+func NewSFServiceInstanceNotFound(name string, err error) *InteroperatorError {
+	return &InteroperatorError{
+		Err:     err,
+		Code:    CodeSFServiceInstanceNotFound,
+		Message: fmt.Sprintf("SFServiceInstance %s not found", name),
+	}
+}
+
+// IsSFServiceInstanceNotFound is true if the error indicates the requested service instance is not found.
+func IsSFServiceInstanceNotFound(err error) bool {
```
Comment (amitmalav) [inline interoperator/pkg/errors/errors.go]: We can remove `Is` from the name.  `IsSFServiceInstanceNotFound ` => `SFServiceInstanceNotFound `

File: interoperator/pkg/errors/errors.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+func NewSFServiceNotFound(name string, err error) *InteroperatorError {
+	return &InteroperatorError{
+		Err:     err,
+		Code:    CodeSFServiceNotFound,
+		Message: fmt.Sprintf("SfService %s not found", name),
+	}
+}
+
+// IsSFServiceNotFound is true if the error indicates the requested service is not found.
+func IsSFServiceNotFound(err error) bool {
```
Comment (amitmalav) [inline interoperator/pkg/errors/errors.go]: `IsSFServiceNotFound`  => `SFServiceNotFound `?

File: interoperator/pkg/errors/errors.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+func NewSFPlanNotFound(name string, err error) *InteroperatorError {
+	return &InteroperatorError{
+		Err:     err,
+		Code:    CodeSFPlanNotFound,
+		Message: fmt.Sprintf("SfPlan %s not found", name),
+	}
+}
+
+// IsSFPlanNotFound is true if the error indicates the requested service plan is not found.
+func IsSFPlanNotFound(err error) bool {
```
Comment (amitmalav) [inline interoperator/pkg/errors/errors.go]: `IsSFPlanNotFound ` => `SFPlanNotFound `

File: interoperator/pkg/errors/errors.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+func NewSFServiceBindingNotFound(name string, err error) *InteroperatorError {
+	return &InteroperatorError{
+		Err:     err,
+		Code:    CodeSFServiceBindingNotFound,
+		Message: fmt.Sprintf("SFServiceBinding %s not found", name),
+	}
+}
+
+// IsSFServiceBindingNotFound is true if the error indicates the requested service binding is not found.
+func IsSFServiceBindingNotFound(err error) bool {
```
Comment (amitmalav) [inline interoperator/pkg/errors/errors.go]: `IsSFServiceBindingNotFound ` => `SFServiceBindingNotFound `

File: interoperator/pkg/errors/errors.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+	return ErrorCode(err) == CodeSFServiceBindingNotFound
+}
+
+// IsNotFound is true if the error indicates any not found error
+func IsNotFound(err error) bool {
```
Comment (amitmalav) [inline interoperator/pkg/errors/errors.go]: `IsNotFound ` => `NotFound `? And for all similar error types also if it makes sense

File: interoperator/pkg/controller/sfservicebinding/sfservicebinding_controller.go (hunk 1 above)
Comment (vivekzhere) [reply interoperator/pkg/controller/sfservicebinding/sfservicebinding_controller.go]: This is in the updateUnbindStatus function. This means delete was already triggered. We just have to not fail the delete here

**Pull request merged**